docs/enhancements/dynamic-fetching-of-openshift-versions.md

```diff
@@ -0,0 +1,289 @@
+for each minor release.
+
+#### Installing an Older Version of a z-stream
+
+As a user of assisted-installer UI, I should have the option to browse the
+different available versions that are fetched from upgrade channels.
+
+#### Configuring the Dynamically-loaded Versions of a Stream
+
+As a system administrator for the assisted-installer, I should have the option
+to configure both dynamically-loaded and statically-loaded OpenShift version.
+To do that, I can provide the following environment variables for example:
+```
+RELEASE_IMAGES='[
+  {
+    "openshift_version": "4.12-multi",
+    "cpu_architecture": "multi",
+    "cpu_architectures": [
+        "x86_64",
+        "arm64",
+        "ppc64le",
+        "s390x"
+    ],
+    "url": "quay.io/openshift-release-dev/ocp-release:4.12.0-rc.4-multi",
+    "version": "4.12.0-rc.4"
+  }
+]'
+
+DYNAMIC_OPENSHIFT_RELEASES_CONFIG='[
+  {
+    "openshift_version": "4.11",
+    "upgrade_channels": [
+        {
+          "cpu_architecture": "x86_64",
+          "channels": ["stable"]
+        }
+    ]
+  },
+  {
+    "openshift_version": "4.11",
+    "upgrade_channels": [
+      {
+        "cpu_architecture": "x86_64",
+        "channels": ["candidate", "stable"]
+      },
+      {
+        "cpu_architecture": "arm64",
+        "channels": ["candidate", "stable"]
+      }
+    ]
+  }
+]'
+```
+
+### Risks and Mitigations
+
+We assume here that each x.y RHCOS version is backwards and forwards compatible
+with all x.y.z OpenShift versions.
+
+We have recently seen occasions where the disk space is being exhausted due to
+the assisted-service caching a lot of different openshift installer binaries.
+It is important to make sure we don't make this problem more severe, as this
+enhancement introduces much more versions which may (or may not) cause use of
+them.
+A possible solution here is implementing the cache as a LRU-like cache, making
+sure we evict the old entries when we reach a certain capacity.
+
+We should make sure we don't consume too much memory when fetching releases.
+It's especially critical when the user queries for the different versions, but
+might happen in other cases too.
+
+With the suggested solution we don't have a way to retain versions that were
+removed from the upgrade-channels API. While it's not probable we should keep
+that in mind.
+
+## Design Details
+
+### Open Questions
+
+* When implementing the binaries LRU cache, should we prioritize the latest
+  versions of each y-release? We expect those to be the most common ones, but
+  maybe it's better to let the LRU mechanism determine what are the common
+  versions by itself?
+* How should we limit the amount of versions handled? In case of an enormous
+  amount of version entries returned from the API, can we make sure to mitigate
+  the issue on our side?
+* Should we find a way to retain versions if they get out of the
+  upgrade-channels API, and if so how. DB can be a solution but so some cache
+  implementations.
```

Comment:
Is this not the reasoning for keeping the existing `RELEASE_IMAGES` variable?

Reply:
It's only one edge-case that can be handled by it (but there's a lot more)
Another one is a release that didn't get into any channel yet (happened with 4.12 EC versions at first)

Reply:
Can you add these examples as potential reasons in by previous question about why we're keeping `RELEASE_IMAGES`. Then you can just remove this question since we have a solution for it.

Reply:
Done!